### DISCUSSION TITLE
ChatAgents2 API proposal feedback

### DISCUSSION THREAD

**Discussion by lukka** [microsoft/vscode-discussions#1039] [Extension Development QnA] [answered]
For a Chat Agent provider, there should be an API to resolve an arbitrary variable. For example, 
- how would a chat agent "A" resolve a variable provided by chat agent "B"? 
   - E.g. the [IChatVariableService ](https://github.com/microsoft/vscode-copilot/blob/644ea7a4b6319dfba3af68d4f8a97a161647ec98/src/base/conversation/context/variableService.ts#L18C14-L18C35)should be publicly exposed. 
   -  Hopefully this is an expected use case. If variables are not meant to be shared across agents, it would not be possible for a an extension to provide a useful context for the in-box provided `@workspace` agent for example.
- _intrinsic_  variables should be resolvable as well, e.g., `#selection`, should be used and useful to any chat agent (not just `@workspace`).
   - The [CurrentSelectionResolved ](https://github.com/microsoft/vscode-copilot/blob/765174c57bf55470bfb5edb5a695e31c918388ed/src/base/conversation/context/resolvers/currentSelection.ts#L14C14-L14C39) functionality could be exposed. 
   - For example this is evident in the [chat-agent-sample](https://github.com/microsoft/vscode-extension-samples/tree/main/chat-agent-sample) where if asked `@cat could you please summarize #selection`, the token `[#selection](values:#selection)` is going to be passed as is (and it is not resolved) in the prompt, e.g.:
```[chat fetch] messages: [{"content":"You are a cat! Reply in the voice of a cat, using cat analogies when appropriate.","role":"system"},{"content":"could you please summarize the [#selection](values:selection)","role":"user"}]```

@isidorn @benmcmorran @sinemakinci1 @hnrqbaggio @spebl FYI

**Comment by lukka** [comment]
Actually the variables names along with its resolved values are available in the `ChatAgentRequest.variables` field which is accessible by the `context:ChatAgentRequest` instance provided to the `ChatAgentHandler`. This includes the in box variables such as `#selection`, which is great!
It would be nice to have some helper function to replace the variable placeholders (such as `[name](values:value)` in the prompt with their own actual values.

**Comment by isidorn** [comment]
@lukka thank you for this feedback. Please keep it coming.

@roblourens @jrieken fyi - this feedback is from our C++ extension

**Comment by roblourens** [comment, marked as answer]
Yeah, the variable values are passed but it is up to the agent to decide how to fit them into the prompt. There are different ways that one could do this and we don't want to be opinionated about it, but we should probably have a sample for that.
